### DISCUSSION TITLE
Deploying multiple functions in dotnet within same csproj

### DISCUSSION THREAD

**Discussion by mdg215199** [kedacore/keda#1897] [Q&A / Need Help] [answered]
I might be missing an obvious, but is there a way to deploy a csproj that has multiple functions within it?  I tried setting up different triggers, but was no luck (again, likely i'm doing something silly).  If I made two projects, I was able to deploy (and no issue with setting up nginx routing to go to right function/triggers).  

Also, if using the azure template with dotnet http trigger with openapi, is there a way to also deploy it with the /api/swagger/ui too?

Thanks in advance for any tips/tricks/info.

**Comment by tomkerkhove** [comment, marked as answer]
When you deploy multiple functions as part of the same project, they are bundled into one container. That means, you'll be scaling them together - But this is definitely supported.

When you say that you had no luck, what were you seeing and how did you deploy it?

> Also, if using the azure template with dotnet http trigger with openapi, is there a way to also deploy it with the /api/swagger/ui too?

This is an Azure Functions question, not really related to KEDA but yes you can do that - https://github.com/Azure/azure-functions-openapi-extension

**Reply by mdg215199** [reply]
> When you deploy multiple functions as part of the same project, they are bundled into one container. That means, you'll be scaling them together - But this is definitely supported.

That was my understanding as well.  That said, I was doing something stupid, and I got that part resolved.  I was too focused on one avenue and you just need to get a good nights sleep.

On the openapi-extension, I got it working locally as well as deploying inside azure functions as well.  When deployed to keda, no luck just yet, but it might have an issue with what i'm doing with nginx.  I'm going to go ahead and close this issue.  If i'm still having issues with getting openapi to show when deployed I will open a QA as well as all code, deploy, yml, etc.   

Also, thanks for your quick response!

**Reply by tomkerkhove** [reply]
Happy to help.

In terms of OpenAPI; KEDA has nothing to do with that so you'll have to look in some other place, I'm sorry.

**Reply by mdg215199** [reply]
Yeah, I assumed that as much.  Though, I'm guessing I would likely get told that since me deploying the docker image to azure function works and it works (which if does) that I likely need to ask any help within the keda community.  I assumed that I was doing something wrong with keda, and likely am still doing something wrong (which is where I asked as a qa).  I'll try to use it with istio directly real quick, as using it already from its gateway as was wanting to keep our functions and dapr by it's own aks (already in production in istio, and wanted to isolate this work so that it doesnt impact dev/stg/uat/prod ci though I did want to test and ensure we could use our gateway with it), and from there using istio gateway with service entry to add it as external endpoint to the mesh.

Anyways, was mainly posting here out of curiosity to see if anyone else had a tip.   Got side tracked on other priorities last couple days, but when I find the issues I'll post the project in case it can help anyone else out (minus the istio part, I'm testing calling the api to keda directly as well as calling through my istio gateway as well

**Reply by tomkerkhove** [reply]
No problem at all!

Feel free to:
- Let us know if you are using it in production and want to be a listed user 
- Let me know what your full app deployment looks like to see if I can help. When you say it worked, was that locally or the same container image?

**Reply by mdg215199** [reply]
> Let us know if you are using it in production and want to be a listed user

For sure.  We are looking at keda as a way to extend our workflow UI to allow customers to create functions within our workflow engine.    I'm pretty excited on the possibilities with keda, and really impressed with running it locally (now, it's deployed to aks w/in rnd/dev).  Previously, we had about 16 micro-services in azure functions within production, though migrated them to aks in order to have a bit more control (designed/ensured the architect could be deployed to both app service and/or azure func prior to deploying it, and thankfully, migrating to aks was fairly smooth).  So, I had considered openfaas, knative, etc. but since we already have experience using azure functions (as well as the documentation around it) keda seemed like a no brainer.  

So, will reach out to you once we are in production as well as more details/website/documentation, as would be great to be on the list.

> Let me know what your full app deployment looks like to see if I can help. When you say it worked, was that locally or the same container image?

I was deploying it to an AKS running k8s 1.19.11 with two nodes running Standard_E4as_v4 (we have several reserved instances, one of which is for rnd).  I'll test it a. bit more, but if i'm still having issues i might reach out to you.  Honestly, everything seems to be working great other than the openapi ui not showing up.  

Thanks for the quick response and willingness for help/knowledge!  

Best regards,
Matt

**Reply by tomkerkhove** [reply]
Sounds great, thank you!

> Honestly, everything seems to be working great other than the openapi ui not showing up.

Make sure that the code allows you to do so. If you say that it workes outside of Kubernetes, is that on dev box or containerized as well? Often OpenAPI specs are not allowed when building in release which could clarify your scenario.
